store/src/pmmr.rs

```diff
@@ -322,82 +312,102 @@ where
-				None
-			}
-		});
-		self.data_file
-			.save_prune(tmp_prune_file_data.clone(), to_rm, record_len)?;
-
-		// 3. update the prune list and save it in place
-		for &(rm_pos, idx) in &self.rm_log.removed[..] {
-			if idx < cutoff_index {
-				self.pruned_nodes.add(rm_pos);
+
+		// Pos we want to get rid of.
+		let rm_pre_cutoff = self.rm_log.removed_pre_cutoff(cutoff_index);
+		let pos_to_rm = removed_excl_roots(rm_pre_cutoff.clone());
+		let leaf_pos_to_rm = removed_leaves(pos_to_rm.clone());
+
+		// 1. Save compact copy of the hash file, skipping removed data.
+		{
+			let record_len = 32;
+
+			let off_to_rm = pos_to_rm
+				.iter()
+				.map(|&pos| {
+					let shift = self.pruned_nodes.get_shift(pos);
+					(pos - 1 - shift.unwrap()) * record_len
+				})
+				.collect();
+
+			self.hash_file
+				.save_prune(tmp_prune_file_hash.clone(), off_to_rm, record_len)?;
+		}
+
+		// 2. Save compact copy of the data file, skipping removed leaves.
+		{
+			let record_len = T::len() as u64;
+
+			let off_to_rm = leaf_pos_to_rm
+				.iter()
+				.map(|pos| {
+					let shift = self.pruned_nodes.get_leaf_shift(*pos);
+					(pos - 1 - shift.unwrap()) * record_len
+				})
+				.collect::<Vec<_>>();
+
+			self.data_file
+				.save_prune(tmp_prune_file_data.clone(), off_to_rm, record_len)?;
+		}
+
+		// 3. Update the prune list and save it in place.
+		{
+			for &pos in &rm_pre_cutoff {
+				self.pruned_nodes.add(pos);
 			}
+
+			write_vec(
+				format!("{}/{}", self.data_dir, PMMR_PRUNED_FILE),
+				&self.pruned_nodes.pruned_nodes,
+			)?;
 		}
-		write_vec(
-			format!("{}/{}", self.data_dir, PMMR_PRUNED_FILE),
-			&self.pruned_nodes.pruned_nodes,
-		)?;
 
-		// 4. move the compact copy of hashes to the hashsum file and re-open it
+		// 4. Rename the compact copy of hash file and reopen it.
 		fs::rename(
 			tmp_prune_file_hash.clone(),
 			format!("{}/{}", self.data_dir, PMMR_HASH_FILE),
 		)?;
 		self.hash_file = AppendOnlyFile::open(format!("{}/{}", self.data_dir, PMMR_HASH_FILE), 0)?;
 
-		// 5. and the same with the data file
+		// 5. Rename the compact copy of the data file and reopen it.
 		fs::rename(
 			tmp_prune_file_data.clone(),
 			format!("{}/{}", self.data_dir, PMMR_DATA_FILE),
 		)?;
 		self.data_file = AppendOnlyFile::open(format!("{}/{}", self.data_dir, PMMR_DATA_FILE), 0)?;
 
-		// 6. truncate the rm log
-		self.rm_log.removed = self.rm_log
+		// 6. Truncate the rm log based on pos removed.
+		// Excluding roots which remain in rm log.
+		self.rm_log
 			.removed
-			.iter()
-			.filter(|&&(_, idx)| idx >= cutoff_index)
-			.map(|x| *x)
-			.collect();
+			.retain(|&(pos, _)| !pos_to_rm.contains(&&pos));
 		self.rm_log.flush()?;
 
 		Ok(())
 	}
 }
+
+/// Filter remove list to exclude roots.
+/// We want to keep roots around so we have hashes for Merkle proofs.
+fn removed_excl_roots(removed: Vec<u64>) -> Vec<u64> {
+	removed
+		.iter()
+		.filter(|&pos| {
+			let (parent_pos, _, _) = family(*pos);
+			removed.contains(&parent_pos)
```

Comment:
I think `removed` is ordered, so `binary_search` will behave a lot better than `contains` here. As the `Vec` could grow large, it's better to keep an eye on O(n^2) operations.

Reply:
Yes it is - good catch on this. Will update.